socorro/external/postgresql/connection_context.py

```diff
@@ -2,15 +2,32 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
+import os
 import socket
 import contextlib
 import psycopg2
 import psycopg2.extensions
+from urlparse import urlparse
 
-from configman.config_manager import RequiredConfig
-from configman import Namespace
+from configman import RequiredConfig, Namespace
+from configman.memoize import memoize
 
+#------------------------------------------------------------------------------
+# we're going to end up parsing the same url several times, there's no need to
+# actually do it more than once, so we wrap the method with a memoizer to speed
+# things along.
+@memoize()
```

Comment:
I sort of see what's going on now. There are two instances of it being "used" repeatedly. 

My original complaint (I most likely mishandled my comments) was about the DRY violation of using `urlparse.urlparse()` logic spread in multiple places with miniscule difference. 

Now I see what you're doing with the `required_config.add_option()` as class attributeS repeatedly. That I don't mind being repeated. The "re-parsing" overhead is negligible. 

I think what I don't like about all of these `default_from_url()` is that when my eyes are on them I don't know what this "url" is. The `default_from_url` isn't a "tight" function. It depends on something globally outside of its internal scope. 

You know me, I like bikeshedding and high quality code that is dumb and easy to understand. 

Solution 1)

```python
from utils import parse_database_url

db_defaults = {
    'hostname': 'localhost', 
    'name': 'breakpad', 
    ...
}
if os.environ.get('database_url'):
    # oh! much better defaults are available
    db_defaults.update(parse_database_url(os.environ['database_url')))

class ConnectionContext(...):
    required_config = Namespace()
    required_config.add_option(
        name='database_hostname',
        default=db_defaults['hostname'],
        doc='the hostname of the database',
        reference_value_from='resource.postgresql'
    )
    ...etc....
```

Solution 2)

Get rid of ALL explosion of database credentials everywhere. ONLY allow for database URLs to configure things ever. E.g. disallow the use of relying on all other defaults but deliberately overriding the password with `setupdb_app.py --database_password=imspecial`.

I'm guessing this is harder to do due to legacy but you will know much better than I. 

Solution 3)

Ignore everything I say because things will be replaced in a month or two anyway. :)